### ISSUE TITLE
Document Baseline plugin

### ISSUE THREAD

**Issue by ryo-manba** [stylelint/stylelint#8391]
### What is the problem you're trying to solve?

Ensure that CSS features follow Baseline support levels.

### What solution would you like to see?

Suggested blueprint:  

- Name: `baseline-no-unsupported`
- Description: Enforce CSS features based on Baseline support.  
- Message: `"Unexpected "${feature}", not a "${availability}" baseline feature`
- Primary option: `true`  
- Secondary options:
  - `available: "widely" | "newly"` (default: `"widely"`)  
- Category: Avoid errors > Unknown
- Included in sharable configs? Yes
- Expanded description:
  [Baseline](https://web-platform-dx.github.io/web-features/) defines a set of CSS features that are consistently supported across major browsers.
  This rule checks whether properties, at-rules, values, and functions comply with the selected Baseline level and flags unsupported features.

For examples: 
```json
{
  "rules": {
    "baseline-no-unsupported": [true, { "available": "widely" }]
  }
}
```  

Managing Baseline data is a key challenge.  
- Like `eslint/css`, the rule will likely need to retrieve data from [web-features](https://npmjs.com/package/web-features).
- Should this data be included within the Stylelint repository, or should it be managed separately in a dedicated repository?  

Prior Art: 
[@eslint/css`require-baseline` rule](https://github.com/eslint/css/pull/33)

### Plugin

I have published a plugin for this rule to demonstrate its potential use and value. 
Feel free to try it out:
🔗 [stylelint-plugin-use-baseline - npm](https://www.npmjs.com/package/stylelint-plugin-use-baseline)

**Comment by ybiquitous** [comment]
@ryo-manba Thanks for the proposal. This idea sounds great, but some of users may not be pleasant to increase the heavy dependency [`web-features`](https://www.npmjs.com/package/web-features) for just one rule. 

So first, I'd recommend to create a plugin to demonstrate its worth. If many people acknowledge the plugin, it might be imported as a built-in rule.

**Comment by jeddy3** [comment]
Starting as a plugin is a great idea.

It'll be good to learn if authoring to a particular baseline is a typical use case. I wonder if most people ignore it in favour of relying on syntax lowering tools like [postcss-preset-env](https://preset-env.cssdb.org/) and [Lightning CSS](https://lightningcss.dev/), or combine baseline and syntax lowering in some way. For anyone interested in this rule, please share your use case so that we can get a fuller picture. 

> Like eslint/css, the rule will likely need to retrieve data from [web-features](https://npmjs.com/package/web-features).

We'll likely need to process the data as a build step to reduce the bundle size, as they do. 

> This rule checks whether properties, at-rules, values, and functions comply with the selected Baseline level and flags unsupported features.

It's probably worth waiting until they work the quirks out in processing the data and then referencing their implementation.

We typically create rules targeting specific constructs, which has helped keep our rule code manageable, .e.g. one for properties, another for functions and so on. Each rule can then have specific options for progressive enhancement appropriate for the construct.

```js
{
  "rules":
    "plugin/function-no-above-baseline": [true, {
      "available": "widely" 
      "ignore": [
        "after-duplicate-property"
      ]
   }],
"plugin/selector-no-above-baseline": [true, {
      "available": "widely" 
      "ignore": [
        "in-at-supports"
      ]
   }]
}
```

Setting the Baseline level could happen in the `languageOptions` configuration property.

(See https://github.com/stylelint/stylelint/issues/8291#issuecomment-2587132762)

**Comment by jeddy3** [comment]
There are some additional use cases outlined in https://github.com/eslint/css/issues/78#issuecomment-2715755220.

**Comment by ryo-manba** [comment]
@ybiquitous 
I followed your suggestion and published a plugin!
🔗 [stylelint-plugin-require-baseline - npm](https://www.npmjs.com/package/stylelint-plugin-require-baseline)

I based the interface on eslint/css’s [require-baseline](https://github.com/eslint/css/blob/9ddd6791d2f559cdc36c27350818db7f1986b5a1/docs/rules/require-baseline.md?plain=1).

**Comment by Mouvedia** [comment]
@ryo-manba you might be interested by https://github.com/eslint/css/blob/main/docs/rules/require-baseline.md?plain=1#L92

**Comment by ybiquitous** [comment]
Great! 👍 

Besides, it seems better to introduce the new plugin in Awesome Stylelint ☺️

**Comment by rviscomi** [comment]
Hi, I contributed to the ESLint rule and just wanted to chime in to say that I'm totally supportive of adding a similar rule to Stylelint and would be happy to contribute in any way. Also, if support is added soon, we might be able to include it in our Baseline tooling announcements at Google I/O next month!

**Comment by jeddy3** [comment]
@rviscomi Thank you for chiming in and offering to contribute. If you have time, would you mind looking over @ryo-manba's [plugin](https://github.com/ryo-manba/stylelint-plugin-require-baseline), which is based on your ESLint rule, to see if any improvements can be made?

> Also, if support is added soon, we might be able to include it in our Baseline tooling announcements at Google I/O next month!

It'd be fab if you could announce that there is a plugin for Stylelint at Google I/O. We consider plugins an essential part of the Stylelint ecosystem and link to a set of officially approved ones [on our website](https://stylelint.io/awesome-stylelint). The Baseline plugin sits alongside one for the older caniuse approach in the [browser compatibility section](https://stylelint.io/awesome-stylelint#browser-compatibility). 

(Because we typically favour creating plugins over adding built-in rules, especially when external datasets are used, this may remain a plugin, at least for some time yet.)

**Comment by rviscomi** [comment]
Sure, I see that there are a couple of open feature requests, and I've filed another one to align it with the updated name `use-baseline`. I can take a look at contributing a few PRs to close those gaps with ESLint.

It's very good to know that this is an officially supported plugin! I'll ask if that's enough for a shoutout at I/O. I'd expect built-in support to help a lot more with adoption and showing Stylelint's commitment, so I'm happy to see that you're still open to it. Is there anything I can do to help, maybe finding more people to share their use cases?

> (Because we typically favour creating plugins over adding built-in rules, especially when external datasets are used, this may remain a plugin, at least for some time yet.)

Does the external dataset concern still apply when the data is preprocessed, as this plugin does?

**Comment by rviscomi** [comment]
Update: I've coordinated with @ryo-manba to add a couple of new features to the plugin.

@jeddy3 let me know if there's anything I can do to help build the case to make this a built-in rule.

**Comment by jeddy3** [comment]
@rviscomi Thanks for the update, and for coordinating with @ryo-manba to add new features to the plugin.

> let me know if there's anything I can do to help build the case to make this a built-in rule.

Thank you for the offer. Your question prompted me to give a lot of thought to how we split plugins and built-in rules, and how that's changed over the years. I'll record my trip down memory lane to help us make a decision, as we don't have the full picture documented anywhere. 

Back in 2017, [we removed several built-in rules](https://stylelint.io/CHANGELOG#800---2017-07-16) so that the community could republish them as plugins, including the `no-unsupported-browser-features` rule (which used the _caniuse_ db under the hood). Although we removed some of these rules due to their weight (which is not an issue for Baseline, as the data can be preprocessed to a smaller size), it marked a point where our approach to built-in rules diverged from ESLint's for other reasons.

In addition to the [criteria in the developer docs](https://stylelint.io/developer-guide/rules#define-the-rule), the [100 or so built-in rules](https://stylelint.io/user-guide/rules) became:

- non-advisory, i.e. don't have potential false positives because the problem can be mitigated in some other way
- tied to a feature in the spec, rather than how a feature is authored (by users) or implemented (by browser vendors)

The former is why rules related to a11y are plugins, despite their usefulness. When people turn on a built-in rule, they can safely assume they are aiming for no errors because the built-in rules have no false positives.

And the latter allows teams to define their own best practices at the configuration level, as illustrated by `time-no-imperceptible` becoming [`time-min-milliseconds`](https://stylelint.io/user-guide/rules/time-min-milliseconds), and is why we [emphasise customising Stylelint](https://stylelint.io/user-guide/customize) in our user guide. It also marked (with the removal of rules like `no-browser-hacks`) a stark departure from CSSLint's approach, where [many of its rules](https://github.com/CSSLint/csslint/wiki/Rules#compatibility) were tied to the best practices of the day.

These two characteristics led to the stability and consistent behaviour of our built-in rules. In contrast, plugins have come and gone over the years as the popularity of the methodology or practice they support has waned.

With this in mind, and if we still think there is value in this dividing line, we may want to keep the Baseline rule as a plugin. 

Like other plugins, it's useful and will benefit many people. It'll "help you narrow down the features that you'll need to use with more care" (https://github.com/web-platform-dx/web-features/issues/2758#issuecomment-2712607492). As a plugin, it can carry a [similar recommendation to the `stylelint-no-unsupported-browser-features` plugin](https://www.npmjs.com/package/stylelint-no-unsupported-browser-features#user-content-recommendations) regarding severity so that people are explicitly aware of the advisory nature of the rule and how they may need to treat it differently from how they do the built-in rules.

Whether it's a built-in rule or a plugin doesn't indicate to our users its value or usefulness. And it's worth reiterating that plugins are an essential part of the Stylelint ecosystem that:

- have a [list of approved ones](https://stylelint.io/awesome-stylelint) on our website
- are documented on the [2nd page](https://stylelint.io/user-guide/customize#custom-rules) of our [user guide](https://stylelint.io/user-guide/get-started)

I appreciate that there's a lot to dig into above, despite my attempts to keep it high-level. Diving into this was a helpful exercise, as it reminded me of some of our past design decisions that I had long since forgotten! 

Let's see if any of the other contributors have thoughts on how we divide plugins and built-in rules. Do we want to consider changing our approach going forward, or does sticking with the status quo benefit our users? If it's the latter, I can update the criteria in our docs so that we have it to hand for next time.

**Comment by jeddy3** [comment]
A quick follow-up to this... if we do decide to keep it as a plugin, we can mention it in the [user guide](https://stylelint.io/user-guide/customize#custom-rules) for better visibility. 

```diff md
- Custom rules do all sorts; from enforcing [strict BEM conventions](https://www.npmjs.com/package/stylelint-selector-bem-pattern) to [strict scales for numeric values](https://www.npmjs.com/package/stylelint-scales).
+ Custom rules do all sorts; from [helping you use Baseline effectively](https://www.npmjs.com/package/stylelint-plugin-use-baseline) to [enforcing strict scales for numeric values](https://www.npmjs.com/package/stylelint-scales).
```

It'll be an opportunity to replace a plugin supporting an older practice (writing BEM conventions by hand) with something more modern (using Baseline to gauge which styles may need to be used with more care).

**Comment by romainmenke** [comment]
The documentation was added in 17.0.0
